### PR TITLE
[docker] Support non-root container

### LINKED ISSUE #10786: [docker] Support non-root container
### Describe your feature request

The Docker image created by `build-docker.sh` requires the root user to run.
Support non-root container is important to have better security. see

  - https://docs.docker.com/engine/security/security/
  - https://engineering.bitnami.com/articles/why-non-root-containers-are-important-for-security.html

Currently, docker image installs python conda environment and libraries under `/root` directory, which non-root user doesn't have permission. Also, [some source code](https://github.com/ray-project/ray/blob/ray-0.8.7/python/ray/autoscaler/command_runner.py#L167) creates files under `/root` directory.

How about create a non-root user 'ray' in the docker image and use `/home/ray` for python conda environment under as well as file upload target directory?

### PR DESCRIPTION
**Pull request by Leemoonsoo**
## Why are these changes needed?

This change makes non-root container images to be created.

Summary of changes

 - `docker/base-deps/Dockerfile` - create 'ray' user and install 'sudo' command
 - `docker/base-deps|ray-deps|ray|ray-ml/Dockerfile` - Dockerfile switches user to ray and use 'sudo' when command need to run with root permission
 - update python source codes that references `/root` dir (to `/home/ray` dir)

## Related issue number

Closes https://github.com/ray-project/ray/issues/10786

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ijrsvt — COMMENTED** [review]
This looks great so far, the only thing thing is that this would break clusters that use the Autoscaler's built-in `Docker` support.  There is currently a dependency on `/root` being the "home directory" in `def docker_start_cmds(...)` in `docker.py`

I think you can get around this by adding an argument of `home_directory` to `docker_start_cmds`. Then in `command_runner::DockerCommandRunner::run_init`, you can do something like
```
        image_env = self.ssh_command_runner.run(
            "docker inspect -f '{{json .Config.Env}}' " + image,
            with_output=True).decode().strip()
        home_directory = "/root"
        for env_var in json.loads(image_env):
            if env_var.startswith("HOME="):
                home_directory = env_var.split("HOME=")[1]
                break
```

**Comment by Leemoonsoo** [comment]
@ijrsvt Thanks for the quick review. Let me ping you again after address `Docker` support.

**Comment by Leemoonsoo** [comment]
@ijrsvt addressed the comment. Please take a look.

File: python/ray/autoscaler/_private/command_runner.py [hunk 1]
```diff
@@ -32,6 +32,7 @@
 HASH_MAX_LENGTH = 10
 KUBECTL_RSYNC = os.path.join(
     os.path.dirname(os.path.abspath(__file__)), "kubernetes/kubectl-rsync.sh")
+RAY_HOME = "/home/ray"  # ray home path in the container image
```
Comment (ijrsvt) [inline python/ray/autoscaler/_private/command_runner.py:35]: hmm, can you move this to `_private/constants.py`?

**Review by ijrsvt — CHANGES_REQUESTED** [review]
1 comment above and it looks good to me. I was wondering if you had done any testing of this? (Minimally just checking that `ray up` works with new and old images?

**Comment by ijrsvt** [comment]
cc @yiranwang52  any concerns with this?

File: python/ray/autoscaler/_private/command_runner.py (hunk 1 above)
Comment (Leemoonsoo) [reply python/ray/autoscaler/_private/command_runner.py]: Moved!

**Comment by Leemoonsoo** [comment]
> 1 comment above and it looks good to me. I was wondering if you had done any testing of this? (Minimally just checking that ray up works with new and old images?

I tested `ray up` works with both old images and new images, using configurations like

```
cluster_name: default
min_workers: 1
max_workers: 1
docker:
    image: "rayproject/ray:latest"
    # image: "rayproject/ray:nightly"
    container_name: "rayc"
    pull_before_run: false
target_utilization_fraction: 0.8
idle_timeout_minutes: 5
provider:
    type: local
    head_ip: "my local host name"
    worker_ips: []
auth:
    ssh_user: "my local host user"
    ssh_private_key: ~/.ssh/id_rsa
head_node: {}
worker_nodes: {}
setup_commands: []
head_setup_commands: []
worker_setup_commands: []
setup_commands:
    - cd $HOME && test -e ray || git clone https://github.com/ray-project/ray.git
    - cd $HOME/ray && git fetch
#    - cd $HOME/ray/python && pip install -e .
head_start_ray_commands:
    - ray stop
    - ulimit -c unlimited && ray start --head --port=6379 --autoscaling-config=~/ray_bootstrap_config.yaml
worker_start_ray_commands:
    - ray stop
    - ray start --address=$RAY_HEAD_IP:6379
```

Let me know any other things you want to verify :)

**Comment by ijrsvt** [comment]
@Leemoonsoo  That looks  like a great setup -- the only other things to check are 
* K8s
* file_mounts on Docker/k8s

Thanks again for doing this!!

**Comment by ijrsvt** [comment]
@Leemoonsoo  It looks like this PR is failing some autoscaler unit tests. I think the failure has to do with not adding a `MockProcessRunner::respond_to_call` for `"docker inspect -f '{{json .Config.Env}}'`. 

For an example of what to do, `test_autoscaler::AutoscalingTest::testGetOrCreateHeadNode` is a great example. Let me know if you have any problems fixing the tests!

**Comment by Leemoonsoo** [comment]
Okay. Let me try to fix unit tests as well.

**Comment by Leemoonsoo** [comment]
@ijrsvt I updated tests. Could you have a quick look and check if I'm in the right direction?

**Comment by ijrsvt** [comment]
@Leemoonsoo Yep! That looks great! I think possibly merge with master and then it should be good! One quick question--can you get away with a smaller value than `100` in the : ` ["[]" for _  in range(100)]`?

**Comment by richardliaw** [comment]
@Leemoonsoo let us know if you need any help! This is an exciting PR :)

**Review by ijrsvt — APPROVED** [review]
Looks good!

**Pull request merged**